addons/marketing_campaign/models/marketing_campaign_activity.py

```diff
@@ -0,0 +1,89 @@
+- Custom Action: execute a predefined action, e.g. to modify the fields of the resource record
+""")
+    email_template_id = fields.Many2one('mail.template', string='Email Template',
+                                        help='The email to send when this activity is activated')
+    report_id = fields.Many2one('ir.actions.report.xml', string='Report',
+                                help='The report to generate when this activity is activated', )
+
+    server_action_id = fields.Many2one('ir.actions.server', string='Action',
+                                       help="The action to perform when this activity is activated")
+    to_ids = fields.One2many('marketing.campaign.transition',
+                             inverse_name='activity_from_id',
+                             string='Next Activities')
+    from_ids = fields.One2many('marketing.campaign.transition',
+                               inverse_name='activity_to_id',
+                               string='Previous Activities')
+    variable_cost = fields.Float(help="Set a variable cost if you consider that every campaign item that has reached this point has entailed a certain cost. You can get cost statistics in the Reporting section",
+                                 digits=dp.get_precision('Product Price'))
+    revenue = fields.Float(help="Set an expected revenue if you consider that every campaign item that has reached this point has generated a certain revenue. You can get revenue statistics in the Reporting section",
+                           digits=dp.get_precision('Account'))
+    signal = fields.Char(
+        help='An activity with a signal can be called programmatically. Be careful, the workitem is always created when a signal is sent')
+    keep_if_condition_not_met = fields.Boolean(string="Don't Delete Workitems",
+                                               help="By activating this option, workitems that aren't executed because the condition is not met are marked as cancelled instead of being deleted.")
+
+    @api.model
+    def search(self, args, offset=0, limit=None, order=None, count=False):
+        segment_id = self.env.context.get('segment_id')
+        if segment_id:
+            segment = self.env['marketing.campaign.segment'].browse(segment_id)
+            return segment.campaign_id.activity_ids.ids
+        return super(MarketingCampaignActivity, self).search(args=args, offset=offset, limit=limit, order=order, count=count)
+
+    def _get_action_types(self):
+        return [
+            ('email', 'Email'),
+            ('report', 'Report'),
+            ('action', 'Custom Action'),
+            # TODO implement the subcampaigns.
+            # TODO implement the subcampaign out. disallow out transitions from
+            # subcampaign activities ?
+            #('subcampaign', 'Sub-Campaign'),
+        ]
+
+    @api.multi
+    def _process_wi_email(self, activity, workitem):
+        return activity.email_template_id.send_mail(workitem.res_id)
```

Comment:
self and activity cannnot be same ? is it required to pass activity as argument ?